From e0dd2594bcd165eb3c8b62b1bbe2b229e39bff92 Mon Sep 17 00:00:00 2001
From: Jay Logue <jay.logue@gmail.com>
Date: Fri, 24 Apr 2026 13:01:32 -0700
Subject: [PATCH] =?UTF-8?q?=EF=BB=BFFix=20GPIO=20pin=20export=20race=20con?=
 =?UTF-8?q?dition?=
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

-- Rewrote gpios_c::export_pin() to avoid an apparent race condition
in later kernels when exporting a GPIO pin.

The problem occurred when exporting a GPIO pin that had already been
exported. In modern kernels, this can result in the GPIO pin being
un-exported (due to an internal error), which causes the app to
fail later.

The fix is to skip exporting GPIO pins that are already exported.

In addition to this change, the code in export_pin() was improved to
wait for a period of time for the GPIO directory to appear and to fail
with an appropriate error if it does not.
---
 10.01_base/2_src/arm/gpios.cpp | 33 +++++++++++++++++++++++----------
 1 file changed, 23 insertions(+), 10 deletions(-)

diff --git a/10.01_base/2_src/arm/gpios.cpp b/10.01_base/2_src/arm/gpios.cpp
index 0d602e1..3b1fc64 100644
--- a/10.01_base/2_src/arm/gpios.cpp
+++ b/10.01_base/2_src/arm/gpios.cpp
@@ -33,6 +33,7 @@
 #include <time.h>
 #include <assert.h>
 #include <fcntl.h>
+#include <unistd.h>
 #include <sys/mman.h>
 #include <sys/stat.h>
 
@@ -133,19 +134,31 @@ void gpios_c::export_pin(gpio_config_t *pin)
     char fname[256];
     FILE *f;
     struct stat statbuff;
+    int retry = 0;
 
-    sprintf(fname, "/sys/class/gpio/export");
-    f = fopen(fname, "w");
-    if (!f)
-        FATAL("Can not open %s", fname);
-    fprintf(f, "%d\n", pin->linear_no);
-    fclose(f);
-
-    // verify: now there should appear the control directory for the pin
     sprintf(fname, "/sys/class/gpio/gpio%d", pin->linear_no);
-    if (stat(fname, &statbuff) != 0 || !S_ISDIR(statbuff.st_mode))
-        FATAL("Gpio control dir %s not generated", fname);
 
+    // export GPIO and wait for control directory to appear
+    // do nothing if GPIO already exported
+    while (true) {
+        if (stat(fname, &statbuff) == 0)
+            break;
+        if (errno != ENOENT) {
+            FATAL("Error accessing %s: %d\n", fname, errno);
+        }
+        if (retry == 0) {
+            f = fopen("/sys/class/gpio/export", "w");
+            if (!f)
+                FATAL("Cannot open /sys/class/gpio/export: %d\n", errno);
+            fprintf(f, "%d\n", pin->linear_no);
+            fclose(f);
+        }
+        retry++;
+        if (retry >= 50) {
+            FATAL("Timed out waiting for %s\n", fname);
+        }
+        usleep(10000);  // Wait 10ms between checks
+    }
 }
 
 /* export the NON-PRU pins: */
-- 
2.43.0

